### ISSUE TITLE
Package version?

### ISSUE THREAD

**Issue by zmc** [ceph/shaman#57]
I can't find any reference to package versions in Shaman (or Chacra), but teuthology does need that information.

An example link in gitbuilder: http://gitbuilder.ceph.com/ceph-deb-xenial-x86_64-basic/ref/jewel/version

``teuthology-suite`` (I think) only uses it as an indicator that the build is complete, but the install task needs it so that it can do things like ``apt get install ceph=10.2.2-501-g12b374f-1xenial``.

**Comment by andrewschoen** [comment]
Ok, I think we can provide this. We also generate a build artifact that contains the version:

https://jenkins.ceph.com/job/ceph-dev-setup/lastSuccessfulBuild/artifact/dist/version/*view*/

What we want to do is build a way to send version (and other build information) to chacra as JSON metadata stored on a repo object. That build information would then be sent to shaman and exposed in the metadata here as well for each repo.

**Comment by alfredodeza** [comment]
A `version` key is now provided by default in the `extra` field. This is an example of a repository object:

    {
    
        "status": "ready",
        "sha1": "9bfc0cf178dc21b0fe33e0ce3b90a18858abaf1b",
        "extra": {
            "node_name": "158.69.92.26+centos7_huge__540a47bf-c1cf-4852-a8e0-e841b5370ddd",
            "build_url": "https://jenkins.ceph.com/job/ceph-dev-build/ARCH=x86_64,AVAILABLE_ARCH=x86_64,AVAILABLE_DIST=centos7,DIST=centos7,MACHINE_SIZE=huge/67/",
            "root_build_cause": "MANUALTRIGGER",
            "version": "10.2.2-508-g9bfc0cf",
            "job_name": "ceph-dev-build/ARCH=x86_64,AVAILABLE_ARCH=x86_64,AVAILABLE_DIST=centos7,DIST=centos7,MACHINE_SIZE=huge"
        },
        "url": "https://1.chacra.ceph.com/r/ceph/jewel/9bfc0cf178dc21b0fe33e0ce3b90a18858abaf1b/centos/7/flavors/default/",
        "distro_codename": null,
        "modified": "2016-08-31 13:02:14.720608",
        "distro_version": "7",
        "project": "ceph",
        "flavor": "default",
        "ref": "jewel",
        "chacra_url": "https://1.chacra.ceph.com/repos/ceph/jewel/9bfc0cf178dc21b0fe33e0ce3b90a18858abaf1b/centos/7/flavors/default/",
        "archs": [
            "x86_64",
            "source"
        ],
        "distro": "centos"
    
    }

**Comment by zmc** [comment]
Thanks,

There's still an issue here though. Given [this query](https://shaman.ceph.com/api/search/?status=ready&sha1=9bfc0cf178dc21b0fe33e0ce3b90a18858abaf1b&project=ceph&flavor=default&ref=jewel&distros=centos/7/x86_64,ubuntu/16.04/x86_64)

The two builds with ``extra`` have a ``version`` of ``10.2.2-508-g9bfc0cf``

For centos packages, we have a [version](http://gitbuilder.ceph.com/ceph-rpm-centos7-x86_64-basic/sha1/9bfc0cf178dc21b0fe33e0ce3b90a18858abaf1b/version) of ``10.2.2-508.g9bfc0cf`` and package names like ``ceph-10.2.2-508.g9bfc0cf.x86_64.rpm``

For ubuntu packages, we have a [version](http://gitbuilder.ceph.com/ceph-deb-xenial-x86_64-basic/sha1/9bfc0cf178dc21b0fe33e0ce3b90a18858abaf1b/version) of ``10.2.2-508-g9bfc0cf-1xenial`` and package names like ``ceph-base_10.2.2-508-g9bfc0cf-1xenial_amd64.deb``

I think it's important that the API report a version that's compatible with the package managers that will be used.

**Comment by alfredodeza** [comment]
Since we created this field as extendable (no schema) we can add that information. The deb build script knows about that differing version scheme and can provide it.

To keep some consistency, the scripts will keep `version` but will add a `package_manager_version` field to provide the package-manager-specific need.

For RPMs both these fields would be the same, for DEBs it would be slightly different. Following the examples above, we would see this for DEB repositories:

    "version": "10.2.2-508-g9bfc0cf"
    "package_manager_version": "10.2.2-508-g9bfc0cf-1xenial"

For RPM repos:

    "version": "10.2.2-508-g9bfc0cf"
    "package_manager_version": "10.2.2-508-g9bfc0cf"

**Comment by dmick** [comment]
I was relatively sure that rpms also use _ instead of -; be careful here

**Comment by andrewschoen** [comment]
I just opened https://github.com/ceph/ceph-build/pull/452 to include ``package_manager_version`` in ``extra``.

**Comment by alfredodeza** [comment]
@dmick doesn't look like it does here: 

http://gitbuilder.ceph.com/ceph-rpm-centos7-x86_64-basic/sha1/0a717593c0c375dcdd9c74e93de5e82b974f1419/version

Do you have an example where this happens? The build scripts provide this at build time, so we are just making the system pass that value around:

https://github.com/ceph/ceph-build/blob/master/ceph-dev-build/build/build_rpm#L60

Which comes from: https://jenkins.ceph.com/job/ceph-dev-setup/lastSuccessfulBuild/artifact/dist/version/*view*/

**Comment by zmc** [comment]
Almost there:
For this build:
https://2.chacra.ceph.com/r/ceph/jewel/9bfc0cf178dc21b0fe33e0ce3b90a18858abaf1b/centos/7/flavors/default/SRPMS/

**Comment by dmick** [comment]
and this ^ is because '-' is a reserved character in rpm filenames for delimiting the three fields {name}-{version}-{release}.

I said '_' above because I'm pretty sure old Jenkins ceph jobs used '_', although I'm having a really hard time finding that; on the gbs, it's build-ceph-rpm.sh that does this translation

**Comment by dmick** [comment]
Anyway, here's the story today:   rpm filenames from the build process are controlled by rpmbuild, so use what the specfile says for name, version, and release in Name: , Version:, and Release: fields.  On gb, those are set through autobuild-ceph.git's build-ceph-rpm.sh, which translates ``'-'`` in Release to ``'.'``  On Jenkins, the spec file comes from the source archive, built by make-dist in ceph.git, which has also done the same translation.   

The number of times the version file moves, in ceph-build.git, from release/ to dist/ to release/ is ... frustrating.

**Comment by alfredodeza** [comment]
I see the attempt of a fix here as just providing whatever the correct value is for yum to consume (and for teuthology to request)

The current rpm build does not have access to a "clean" `Release` value (and neither the parent jobs), but it does have the spec file, so it will just provide what is in there already to reformat the `$vers` variable.

**Comment by alfredodeza** [comment]
Trying it out in builds before getting a pull request out:

    + sed -i 's/^%setup.*/%setup -q -n %{name}-11.0.0-82-g3ad5b63/' ceph.spec
    ++ grep Release ceph.spec
    ++ sed 's/Release:[ \t]*//g'
    ++ cut -d % -f 1
    + RPM_RELEASE=82.g3ad5b63
    ++ grep Version ceph.spec
    ++ sed 's/Version:[ \t]*//g'
    + RPM_VERSION=11.0.0
    + PACKAGE_MANAGER_VERSION=11.0.0-82.g3ad5b63

**Comment by alfredodeza** [comment]
Seems like we have a winner here: https://shaman.ceph.com/api/repos/ceph/wip-jlayton-statx/7c59b9e96a37ecf8130efec29ec6443d7793473a/centos/7/flavors/default/

    [
    
        {
            "status": "ready",
            "sha1": "7c59b9e96a37ecf8130efec29ec6443d7793473a",
            "extra": {
                "build_url": "https://jenkins.ceph.com/job/ceph-dev-build/ARCH=x86_64,AVAILABLE_ARCH=x86_64,AVAILABLE_DIST=centos7,DIST=centos7,MACHINE_SIZE=huge/323/",
                "root_build_cause": "MANUALTRIGGER",
                "version": "11.0.0-1898-g7c59b9e",
                "node_name": "158.69.69.225+centos7_huge__d6ddecb5-8e99-40b8-96c0-60da3588d443",
                "job_name": "ceph-dev-build/ARCH=x86_64,AVAILABLE_ARCH=x86_64,AVAILABLE_DIST=centos7,DIST=centos7,MACHINE_SIZE=huge",
                "package_manager_version": "11.0.0-1898.g7c59b9e"
            },
            "url": "https://2.chacra.ceph.com/r/ceph/wip-jlayton-statx/7c59b9e96a37ecf8130efec29ec6443d7793473a/centos/7/flavors/default/",
            "distro_codename": null,
            "modified": "2016-09-02 19:10:59.885623",
            "distro_version": "7",
            "project": "ceph",
            "flavor": "default",
            "ref": "wip-jlayton-statx",
            "chacra_url": "https://2.chacra.ceph.com/repos/ceph/wip-jlayton-statx/7c59b9e96a37ecf8130efec29ec6443d7793473a/centos/7/flavors/default/",
            "archs": [
                "x86_64",
                "source"
            ],
            "distro": "centos"
        }
    
    ]

**Comment by zmc** [comment]
Not quite:
In [this run](http://pulpito.ceph.com/zack-2016-09-09_11:51:05-smoke-jewel-distro-basic-vps/), we're consuming this build:
```
{u'archs': [u'x86_64'],
  u'chacra_url': u'https://2.chacra.ceph.com/repos/ceph/jewel/9bfc0cf178dc21b0fe33e0ce3b90a18858abaf1b/ubuntu/xenial/flavors/default/',
  u'distro': u'ubuntu',
  u'distro_codename': u'xenial',
  u'distro_version': u'16.04',
  u'extra': {u'build_url': u'https://jenkins.ceph.com/job/ceph-dev-build/ARCH=x86_64,AVAILABLE_ARCH=x86_64,AVAILABLE_DIST=xenial,DIST=xenial,MACHINE_SIZE=huge/356/',
             u'job_name': u'ceph-dev-build/ARCH=x86_64,AVAILABLE_ARCH=x86_64,AVAILABLE_DIST=xenial,DIST=xenial,MACHINE_SIZE=huge',
             u'node_name': u'158.69.83.148+xenial_trusty_pbuilder_huge__63b7a51b-1b56-4ff4-8cba-204935f29dc4',
             u'package_manager_version': u'10.2.2-508-g9bfc0cf-1xenial',
             u'root_build_cause': u'MANUALTRIGGER',
             u'version': u'10.2.2-508-g9bfc0cf'},
  u'flavor': u'default',
  u'modified': u'2016-09-06 21:43:28.156613',
  u'project': u'ceph',
  u'ref': u'jewel',
  u'sha1': u'9bfc0cf178dc21b0fe33e0ce3b90a18858abaf1b',
  u'status': u'ready',
  u'url': u'https://2.chacra.ceph.com/r/ceph/jewel/9bfc0cf178dc21b0fe33e0ce3b90a18858abaf1b/ubuntu/xenial/flavors/default/'}
```

And getting errors like:
```
E: Version '10.2.2-508-g9bfc0cf-1xenial' for 'ceph' was not found
```

Packages are named e.g. ```ceph-base_10.2.2-508-g9bfc0cf-1_amd64.deb```

**Comment by alfredodeza** [comment]
This looks like an inconsistency in the packaging script. We have actual distro versions in the name of the binaries for formal releases:

http://download.ceph.com/debian-jewel/pool/main/c/ceph/

@dmick do you know if the build scripts need to be adapted here to fix this? I am thinking that is the disconnect (not the provided `"package_manager_version"`)

Of course we can hack this to comply, but I feel that the correct path is to have the codename in the binary

**Comment by zmc** [comment]
For that sha1, [gitbuilder](http://gitbuilder.ceph.com/ceph-deb-xenial-x86_64-basic/sha1/9bfc0cf178dc21b0fe33e0ce3b90a18858abaf1b/pool/main/c/ceph/) uses ```ceph_10.2.2-508-g9bfc0cf-1xenial_amd64.deb```

**Comment by dmick** [comment]
Yeah, those codenames-in-the-release are added to allow multiple codenames to exist in the same repo, which wasn't an issue for CI builds on GB.  Now that we're doing multiple codenames in a repo, the packages will have to have those names that include codenames, yes

**Comment by andrewschoen** [comment]
I believe that https://github.com/ceph/ceph-build/pull/474 will fix the issue described in https://github.com/ceph/shaman/issues/57#issuecomment-246037415

You can see the shaman information for this build at https://shaman.ceph.com/api/repos/ceph/master/b8c203fb64c80fe52a4d17562ddef39371347d42/ubuntu/xenial/flavors/default/ and you can see the filenames contain the distro codename here: https://2.chacra.ceph.com/binaries/ceph/master/b8c203fb64c80fe52a4d17562ddef39371347d42/ubuntu/xenial/x86_64/

**Comment by zmc** [comment]
@andrewschoen, I just noticed that [that repo](https://2.chacra.ceph.com/r/ceph/master/b8c203fb64c80fe52a4d17562ddef39371347d42/ubuntu/xenial/flavors/default/pool/main/c/ceph/) has packages for two different ``sha1``s. Example:
```
ceph-base_11.0.0-2170-g25b5a89-1_amd64.deb         12-Sep-2016 16:04            56460576
ceph-base_11.0.0-2245-gb8c203f-1xenial_amd64.deb   12-Sep-2016 16:06            32966708
```

I wonder what's going on there?

**Comment by andrewschoen** [comment]
@zmc that is curious. Notice the first binary does not include ``xenial``. My guess is that this binary was created incorrectly during testing, but I'll take a look.

Does having both of those packages in a repo cause trouble for teuthology? Not saying we shouldn't look into this, just curious.

**Comment by zmc** [comment]
In this particular case, it wouldn't be a problem since the packages shouldn't be there have a "smaller" version. If it were "larger", it could certainly cause problems.

I'm mostly mentioning that because it seems like the way shaman/chacra were designed this shouldn't ever happen

**Comment by andrewschoen** [comment]
You're right, it shouldn't happen. I'll take a look.

**Comment by andrewschoen** [comment]
@zmc ok, I think I've figured it out. We've got the test chacra nodes configured to combine all deb versions into the same repo. The other sha1 you see, ``25b5a89``, is also on chacra2 https://2.chacra.ceph.com/binaries/ceph/master/25b5a898f124b0b34831584584c027369892d46f/ so it's being combined with ``b8c203f``.

We've just got to tweak some configuration for the test chacra nodes and this shouldn't happen anymore. I've opened https://github.com/ceph/chacra/issues/196 to track that.
